storage/innobase/handler/ha_innodb.cc

```diff
@@ -17990,12 +17990,16 @@ void lock_wait_wsrep_kill(trx_t *bf_trx, ulong thd_id, trx_id_t trx_id)
   {
     bool aborting= false;
     wsrep_thd_LOCK(vthd);
-    if (trx_t *vtrx= thd_to_trx(vthd))
+    trx_t *vtrx= thd_to_trx(vthd);
+    if (vtrx)
     {
       lock_sys.wr_lock(SRW_LOCK_CALL);
       mysql_mutex_lock(&lock_sys.wait_mutex);
       vtrx->mutex_lock();
-      if (vtrx->id == trx_id && vtrx->state == TRX_STATE_ACTIVE)
+      /* victim transaction is either active or prepared, if it has already
+	 proceeded to replication phase */
+      if (vtrx->id == trx_id && (vtrx->state == TRX_STATE_ACTIVE ||
```

Comment:
Can we just abort transaction in PREPARED state as we have already promised to XA leader that we can commit? Leader can naturally decide ABORT still.

Reply:
Transactions in `XA PREPARE` state are supposed to be safe from any forced rollback. The only allowed actions are `XA ROLLBACK` or `XA COMMIT` from the user. If Galera uses two-phase commit for its internal purposes like the binlog replication does, then it could be acceptable to choose such transactions as a victim. But, then we should check the XID to ensure that the transaction indeed is in internal 2-phase commit and that it actually is safe to trigger rollback.

If we request a `XA PREPARE` transaction to roll back, where would this request actually be processed? As far as I can tell, an `XA PREPARE` transaction can never be waiting for any lock. Hence, `innobase_kill_query()` cannot have any impact on it. I suppose that the code that we execute in this function could cause `thd_kill_level()` to return a different value for the transaction. But, where would the `XA PREPARE` transaction check for that flag? I do not see the flag being checked even in `innobase_close_connection()`.

@sjaakola If the problem is resolved even when this change is not present, I would suggest to omit this change.

Reply:
this patch would indeed violate XA transaction processing. Fixed in latest commit by restricting this stage of BF aborting only for non XA transactions preparing for wsrep replication